### ISSUE TITLE
`torch.compile` produces different output for Transformer using `torch.einsum` projections compared to eager mode

### ISSUE THREAD

**Issue by prorrice** [pytorch/pytorch#179575]
### 🐛 Describe the bug

`torch.compile` with `inductor` backend produces numerically different results for a Transformer block that uses `torch.einsum("b i d, b d h -> b i h", ...)` for Q/K/V and output projections with element-wise bias addition, followed by multi-head scaled dot-product attention, LayerNorm, and FFN with residual connections.

Inductor's `replace_einsum_to_pointwise` pass rewrites the batched einsum into optimized bmm/matmul operations. This replacement changes the floating-point accumulation order, and the numerical differences compound through the attention mechanism and residual connections, producing a significant output mismatch.

Eager mode is perfectly deterministic (max_var=0 across runs), confirming this is a systematic computation difference introduced by the Inductor optimization, not GPU non-determinism.

**Note**: The model uses `nn.Parameter` with a batch dimension baked into the weight tensors (e.g., `torch.randn(batch_size, d_model, d_model)`). This is how the fuzzer generated the model — it exercises the batched einsum pattern that triggers `replace_einsum_to_pointwise`.

### Minimal reproducer

```python
import os
os.environ["TRITON_BACKENDS_IN_TREE"] = "1"

import torch
import torch.nn as nn
import torch.nn.functional as F

class MultiHeadAttention(nn.Module):
    def __init__(self, d_model=512, num_heads=8, batch_size=4, seq_len=16):
        super().__init__()
        self.d_model = d_model
        self.num_heads = num_heads
        self.head_dim = d_model // num_heads

        # Batch-specific projection weights (triggers batched einsum pattern)
        self.q_proj = nn.Parameter(torch.randn(batch_size, d_model, d_model) * 0.02)
        self.k_proj = nn.Parameter(torch.randn(batch_size, d_model, d_model) * 0.02)
        self.v_proj = nn.Parameter(torch.randn(batch_size, d_model, d_model) * 0.02)
        self.out_proj = nn.Parameter(torch.randn(batch_size, d_model, d_model) * 0.02)

        # Batch-specific biases
        self.q_bias = nn.Parameter(torch.randn(batch_size, seq_len, d_model) * 0.02)
        self.k_bias = nn.Parameter(torch.randn(batch_size, seq_len, d_model) * 0.02)
        self.v_bias = nn.Parameter(torch.randn(batch_size, seq_len, d_model) * 0.02)
        self.out_bias = nn.Parameter(torch.randn(batch_size, seq_len, d_model) * 0.02)

    def forward(self, x):
        batch_size, seq_len, d_model = x.shape

        # einsum projections + element-wise bias add (target pattern)
        q = torch.einsum("b i d, b d h -> b i h", x, self.q_proj) + self.q_bias[:batch_size, :seq_len, :]
        k = torch.einsum("b i d, b d h -> b i h", x, self.k_proj) + self.k_bias[:batch_size, :seq_len, :]
        v = torch.einsum("b i d, b d h -> b i h", x, self.v_proj) + self.v_bias[:batch_size, :seq_len, :]

        # Reshape to multi-head
        q = q.view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        k = k.view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        v = v.view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)

        # Scaled dot-product attention
        scale = self.head_dim ** 0.5
        attn = torch.matmul(q, k.transpose(-2, -1)) / scale
        attn = F.softmax(attn, dim=-1)
        out = torch.matmul(attn, v)

        # Merge heads and output projection
        out = out.transpose(1, 2).contiguous().view(batch_size, seq_len, d_model)
        out = torch.einsum("b i d, b d h -> b i h", out, self.out_proj) + self.out_bias[:batch_size, :seq_len, :]
        return out


class TransformerBlock(nn.Module):
    def __init__(self, d_model=512, num_heads=8, batch_size=4, seq_len=16):
        super().__init__()
        self.attn = MultiHeadAttention(d_model, num_heads, batch_size, seq_len)
        self.norm1 = nn.LayerNorm(d_model)
        self.norm2 = nn.LayerNorm(d_model)
        self.ffn = nn.Sequential(
            nn.Linear(d_model, d_model * 4),
            nn.ReLU(),
            nn.Linear(d_model * 4, d_model),
        )

    def forward(self, x):
        # Pre-norm residual attention
        x_norm = self.norm1(x)
        attn_out = self.attn(x_norm)
        x = x + attn_out
        # Pre-norm residual FFN
        x = x + self.ffn(self.norm2(x))
        return x


device = "cuda"
torch.manual_seed(42)
model = TransformerBlock(d_model=512, num_heads=8, batch_size=4, seq_len=16).to(device).eval()
x = torch.randn(4, 16, 512, device=device)

# Eager: deterministic
with torch.no_grad():
    ref = model(x)
    ref2 = model(x)
print(f"Eager deterministic: {(ref - ref2).abs().max().item():.6e}")

# Compiled
torch._dynamo.reset()
compiled = torch.compile(model, backend="inductor")
with torch.no_grad():
    comp = compiled(x)

diff = (ref.float() - comp.float()).abs()
print(f"Max diff: {diff.max().item():.6e}")
print(f"Mean diff: {diff.mean().item():.6e}")
print(f"Match (atol=1e-5): {torch.allclose(ref, comp, atol=1e-5, rtol=1e-4)}")
```

### Behavior summary

| Mode | Result | Notes |
|------|--------|-------|
| Eager | Reference output | Perfectly deterministic across runs (max_var=0) |
| `torch.compile(backend="inductor")` | Different output | Numerical differences from einsum→bmm replacement compound through attention + residuals |

### Notes

- Eager mode is perfectly deterministic (max_var=0), ruling out GPU non-determinism.
- Inductor's `replace_einsum_to_pointwise` pass rewrites `einsum('b i d, b d h -> b i h', A, B)` as `torch.bmm(A, B)` or equivalent matmul. The replacement changes accumulation order.
- Numerical differences from the Q/K/V projections compound through the softmax attention mechanism and propagate through two residual connections, amplifying the final output difference.
- The batch-specific `nn.Parameter` weight tensors (shape `[batch_size, d_model, d_model]`) are the precise pattern that triggers the batched einsum optimization path.

### Error logs

outputs silently differ.

### Versions

```
PyTorch version: 2.12.0.dev20260327+cu126
Python: 3.10.12
OS: Ubuntu 22.04.5 LTS (WSL2)
GPU: NVIDIA GeForce RTX 3060 Laptop GPU
CUDA: 12.6
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh @anijain2305

topic: fuzzer

**Comment by eellison** [comment]
@prorrice , bitwise equivalence is not the current expectation of inductor. please, show the actual divergence and  use `torch._dynamo.utils.same`, as per fuzzer guidelines, comparing to fp32 ref.
